walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,61 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.Predicate;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventsList = new LinkedList<>();
+    List<Event> eventsListOptional = new LinkedList<>();
+    for (Event event : events) {
+      // add event to list if event and request have attendees in common
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        eventsList.add(event);
+        eventsListOptional.add(event);
+        // account for optional attendees
+      } else if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+        eventsListOptional.add(event);
+      }
+    }
+    Collection<TimeRange> freeSlots = getFreeSlots(eventsListOptional);
+    Predicate<TimeRange> notEnoughTime = (TimeRange x) -> x.duration() < request.getDuration();
```

Comment:
Suggest naming this to indicate to indicate that it's a predicate / function.

Also, nit, but s\x\tr

(this means substitute)

Reply:
Could you explain further on what you mean by s\x\tr?

Reply:
changed predicate name

Reply:
Haha it's an abbreviation that you may eventually see more of when talking to programmers.  It comes from an actual three part command used in some Linux commands (I think).  The three parts are separated by slashes.
The 's' stands for "substitute" (I think.. the name of the command).
The second part is the string that she wants you to delete.
And the third part stands for what she wants you to replace it with.

So she's saying to substitute "tr" for "x" in the line she commented on.  i.e. use a more descriptive variable than x.  Namely, "tr", standing for TimeRange.

Reply:
ah I see, good to know!